Resolve "Replace liballoc with Custom Heap"

Kfeavel opened the pull request:
Closes #367

Kfeavel committed:
Rename `heap` to `Liballoc`

Moved so that the new heap code can live in Heap.cpp/.hpp

Kfeavel committed:
Refactor LinkedList to not require template
Classes should instead publically inherit the `LinkedListNode` class

Kfeavel committed:
Renamed `LinkedListNode` to `Node`

Kfeavel committed:
Begin writing new heap code

micahswitzer commented on Kernel/Library/string.cpp:
```diff
@@ -138,11 +138,27 @@ void* memmove(void* destptr, const void* srcptr, size_t size)
-    return dstptr;
+    uint8_t *bDst;
+    uint8_t *bSrc;
+    uint32_t *lDst = (uint32_t*)dst;
```
Why not use `size_t`'s here?

Kfeavel replied:
Because that's how liballoc had it. You're right, probably should be size_t's though.

micahswitzer replied:
Yeah and I have no idea if they need to unaligned start/destination addresses, but that might be something to account for seeing as this is indented to be a generic `memcpy` replacement.

micahswitzer commented on Kernel/Memory/Heap.cpp:
```diff
@@ -0,0 +1,199 @@
+
+class Major : public LinkedList::Node {
+public:
+    void initialize(size_t pages)
```
Why isn't this a constructor?

Kfeavel replied:
Because we have to request pages manually (no new or malloc by this point) so we have to manually cast the memory returned into a class.

micahswitzer commented on Kernel/Memory/Heap.cpp:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    // Ensure pages is at least equal to the minumum size;
+    if (pages < pageCount) {
```
It's not clear that `pageCount` is a constant

Kfeavel replied:
How would you denote that?

micahswitzer replied:
Wow the style guide (https://webkit.org/code-style-guidelines/) is surprisingly unclear on that. I guess just mark it `const`.

micahswitzer commented on Kernel/Memory/Heap.cpp:
```diff
@@ -0,0 +1,199 @@
+        panic("Out of memory!");
+    }
+
+    major->initialize(pages);
```
Ah I see why you didn't use a constructor. I think we should be using "placement new" for this kind of thing.

https://stackoverflow.com/questions/222557/what-uses-are-there-for-placement-new

Kfeavel replied:
Oh interesting. I'll look into that.

Kfeavel committed:
Partially complete `malloc` case 2

Kfeavel committed:
Make `pageCount` constant

Kfeavel committed:
Fix potential deadlock when out of memory

Kfeavel committed:
Complete `malloc` through part 4.1
Likely will need fixing in some areas

micahswitzer commented on Kernel/Memory/paging.cpp:
```diff
@@ -226,21 +226,30 @@ static uintptr_t findNextFreeVirtualAddress(size_t seq)
+
+    page = (void*)(free_idx * ARCH_PAGE_SIZE);
+
+fail:
```
This would be a good place to use RAII semantics to release the lock on return.

Kfeavel replied:
Yeah, I was thinking that too. I was probably going to make a lambda function for mutexes and semaphores like we do for interrupt critical regions.

Kfeavel replied:
Actually, what I'll probably do is write a wrapper class around Mutex that uses RAII. I could then write another wrapper class for lambda functionality, that way we have either one we want.

Kfeavel replied:
Resolved by #370

Kfeavel committed:
Complete `malloc` implementation
Still need to fix pointer alignment issue

Kfeavel committed:
Fix heap align compiler warning
Needed to pass pointer of pointer for alignment to work properly

Kfeavel committed:
Use getters and setters for heap Minor
Add getters for heap information (bytes allocated & bytes in use)

Kfeavel committed:
Use RAII lock in `malloc`

Kfeavel committed:
Fix head class naming & use linkedlist in `Major`

Kfeavel committed:
Fix incorrect LinkedList logic and parameter order

Kfeavel committed:
Rename `LinkedList` to `LinkedListUnmanaged`

Kfeavel committed:
Replace `reinterpret_cast` with `static_cast`

Kfeavel committed:
Complete `free` implementation
Still needs testing

Kfeavel committed:
Complete `calloc` implementation

GCC's static analyzer reports a `malloc` pointer leak, but I don't
believe that this is correct. If I had to guess, the fact that we're writing
our own implementation of `malloc` is confusing the analyzer. For future
reference, disable the `-Wno-analyzer-malloc-leak` flag in the kernel
SConscript to see the exact analyzer error

Kfeavel committed:
Complete `realloc` implementation

Kfeavel committed:
Remove `liballoc`

Kfeavel committed:
Rework spinner to stress test heap allocation
Something is freezing around the point of allocating 3952 bytes

Kfeavel committed:
Initialize heap and panic on stress test failure

micahswitzer commented on Kernel/Applications/spinner.cpp:
```diff
@@ -10,19 +10,37 @@
+        if (ptr == nullptr) {
+            panic("null malloc in stress test!");
+        }
+        memset(ptr, 0xCAFEBABE, size / sizeof(0xCAFEBABE));
```
This is only writing one quarter of the memory allocated since `memset` only writes bytes of data

micahswitzer commented on Kernel/Applications/spinner.cpp:
```diff
@@ -10,19 +10,37 @@
     Console::printf("\n");
     int i = 0;
     const char spinnay[] = { '|', '/', '-', '\\' };
+    for (size_t size = 1; size < 1024 * 1024 * 1024; size++) {
```
I'm offended that you didn't just write a new task for this test. Poor `spinner` doesn't deserve to be abused like this.

micahswitzer replied:
It might also be nice to test this code outside of the kernel so that it's easier to debug.

Kfeavel replied:
Reset the spinner in 8230d4fec34dadabde6c8370b30c4787170b2129. Its suffering is over.

Kfeavel committed:
Heap: Fix incorrect `if` condition in `malloc`

Kfeavel committed:
Heap: added log output for debugging

Kfeavel committed:
Heap: Use InsertBack in `malloc` case 4.1

Kfeavel committed:
Heap: Add more debug logging

Kfeavel committed:
Heap: Adding missing call to `unalign` in `free`

Kfeavel committed:
Heap: Fix heap magic value by using `rand` twice

Kfeavel committed:
Lib: Change data types in `memmove` and `memcpy`

Kfeavel committed:
Replace `debugf` with `Logger::Debug`

Kfeavel committed:
Check for duplicate pointers in memory stress test

Kfeavel committed:
Fix infinite loop in heap case 1

Kfeavel committed:
Change `startedBet` to boolean

Kfeavel committed:
Fix minor block not being updated in case 4.1

Kfeavel committed:
Fix backwards linked list insertion arguments

Kfeavel committed:
Demote heap logging from `Debug` to `Trace`

Kfeavel committed:
Reset spinner app

Kfeavel commented:
Will write some unit tests for this soon. Thanks to @micahswitzer for writing a basic one used for testing ([`janky-heap-tests`](https://github.com/XyrisOS/xyris/tree/janky-heap-tests)).

Kfeavel committed:
Remove inconsistent linked list docs

Kfeavel committed:
Add brief to `New.hpp`

Kfeavel committed:
Restructure unit tests

Kfeavel committed:
Add heap unit test

Kfeavel committed:
Exclude unit test code from analysis
Prevents VSCode from becoming confused about class definitions

Kfeavel committed:
Reorganize stubbed unit test code

Kfeavel committed:
Add `__xyris__` and `__kernel__` to VSCode defines

Kfeavel committed:
Cleanup heap unit test

Kfeavel committed:
Change heap byte alignment from 16 to 4

Kfeavel committed:
Copy unit test from `liballoc`
Tests allocations and frees of randomized sizes and positions

Kfeavel committed:
Create VSCode configuration for unit tests
Prevents Intellisense from becoming confused regarding definitions
conflicting or being declared in different namespaces

Kfeavel committed:
Use `std::` on functions when possible
Should help remove ambiguity about what function is being called
(xyris vs host)

Kfeavel committed:
Update `Tests/Logger.cpp` to match kernel
